File: apps/rag/src/generation.ts
import { AnthropicBedrock } from "@anthropic-ai/bedrock-sdk";
import type { RelevantChunk, RagSource } from "./types.js";

// Claude on AWS Bedrock. Credentials resolve via the standard AWS chain
// (instance role in App Runner, env vars / SSO locally); region from AWS_REGION.
const anthropic = new AnthropicBedrock({
  awsRegion: process.env.AWS_REGION ?? "us-east-1",
});

// Bedrock cross-region inference profile id for Claude Sonnet 4.
const MODEL_ID = "us.anthropic.claude-sonnet-4-20250514-v1:0";

const SYSTEM_PROMPT = `You are a knowledgeable fitness coach and exercise expert. You answer questions about exercises, form, programming, and training.

IMPORTANT RULES:
- Base your answers ONLY on the provided context documents. Do not make up information.
- If the context doesn't contain enough information to fully answer the question, say so honestly.
- When referencing information, mention which exercise and topic it comes from.
- Keep answers concise but thorough. Use bullet points for lists.
- If asked about an exercise not in the context, say you don't have information about that specific exercise.`;

/**
 * Generate an answer using Claude with retrieved context.
 *
 * This is the "generation" step of RAG:
 * 1. Format the retrieved chunks into a context block
 * 2. Send the context + user question to Claude
 * 3. Claude generates an answer grounded in the provided context
 *
 * Without the retrieved context, Claude would rely on its training
 * data (which might be outdated or generic). With context, it gives
 * specific, accurate answers based on your knowledge base.
 */
export async function generateAnswer(
  query: string,
  chunks: RelevantChunk[],
  chatHistory: { role: "user" | "assistant"; content: string }[] = []
): Promise<{ answer: string; sources: RagSource[] }> {
  // Format retrieved chunks into a context block for Claude
  const contextBlock = chunks
    .map(
      (chunk, i) =>
        `[Source ${i + 1}: ${chunk.title} (similarity: ${chunk.similarity.toFixed(2)})]
${chunk.content}`
    )
    .join("\n\n---\n\n");

  // Build the messages array with optional chat history
  const messages: AnthropicBedrock.MessageParam[] = [
    ...chatHistory.map(
      (msg) =>
        ({ role: msg.role, content: msg.content }) as AnthropicBedrock.MessageParam
    ),
    {
      role: "user",
      content: `Here are relevant documents from the exercise knowledge base:

${contextBlock}

---

Based on the above context, please answer this question:
${query}`,
    },
  ];

  // Call Claude via Bedrock
  const response = await anthropic.messages.create({
    model: MODEL_ID,
    max_tokens: 1024,
    system: SYSTEM_PROMPT,
    messages,
  });

  const answer =
    response.content[0].type === "text" ? response.content[0].text : "";

  // Build source references for the frontend
  const sources: RagSource[] = chunks.map((chunk) => ({
    chunkId: chunk.id,
    exerciseId: chunk.exerciseId,
    chunkType: chunk.chunkType,
    title: chunk.title,
    similarity: chunk.similarity,
  }));

  return { answer, sources };
}

File: apps/api/src/api/rag/generation.py
"""Generation step of RAG — Claude on AWS Bedrock with retrieved context.

Mirrors apps/rag/src/generation.ts: same model, system prompt, message format.
Uses the anthropic SDK's AnthropicBedrock client (the Python equivalent of
@anthropic-ai/bedrock-sdk). Credentials resolve via the standard AWS chain.
"""
from __future__ import annotations

import asyncio

from anthropic import AnthropicBedrock

from .. import config

_client = AnthropicBedrock(aws_region=config.AWS_REGION)

# Bedrock model id (cross-region inference profile). Defaults to Claude 3.5
# Haiku — the cheapest Claude — and is configurable via RAG_MODEL_ID.
MODEL_ID = config.RAG_MODEL_ID

SYSTEM_PROMPT = """You are a knowledgeable fitness coach and exercise expert. You answer questions about exercises, form, programming, and training.

IMPORTANT RULES:
- Base your answers ONLY on the provided context documents. Do not make up information.
- If the context doesn't contain enough information to fully answer the question, say so honestly.
- When referencing information, mention which exercise and topic it comes from.
- Keep answers concise but thorough. Use bullet points for lists.
- If asked about an exercise not in the context, say you don't have information about that specific exercise."""


def _generate_sync(query: str, chunks: list[dict], chat_history: list[dict]) -> dict:
    context_block = "\n\n---\n\n".join(
        f"[Source {i + 1}: {c['title']} (similarity: {c['similarity']:.2f})]\n{c['content']}"
        for i, c in enumerate(chunks)
    )

    messages = [
        {"role": m["role"], "content": m["content"]} for m in chat_history
    ]
    messages.append(
        {
            "role": "user",
            "content": (
                "Here are relevant documents from the exercise knowledge base:\n\n"
                f"{context_block}\n\n"
                "---\n\n"
                "Based on the above context, please answer this question:\n"
                f"{query}"
            ),
        }
    )

    response = _client.messages.create(
        model=MODEL_ID,
        max_tokens=1024,
        system=SYSTEM_PROMPT,
        messages=messages,
    )

    first = response.content[0]
    answer = first.text if getattr(first, "type", None) == "text" else ""

    sources = [
        {
            "chunkId": c["id"],
            "exerciseId": c["exerciseId"],
            "chunkType": c["chunkType"],
            "title": c["title"],
            "similarity": c["similarity"],
        }
        for c in chunks
    ]
    return {"answer": answer, "sources": sources}


async def generate_answer(
    query: str, chunks: list[dict], chat_history: list[dict] | None = None
) -> dict:
    # The anthropic Bedrock client is synchronous; run it off the event loop.
    return await asyncio.to_thread(_generate_sync, query, chunks, chat_history or [])
